### ISSUE TITLE
OpenSSL ossl streamDriver authMode does not allow clients to connect without client certifcates

### ISSUE THREAD

**Issue by JPvRiel** [rsyslog/rsyslog#3837]
### Related

[openssl requires mutual ssl auth even when mode anon is configured](https://github.com/rsyslog/rsyslog/issues/3037) was closed based on [ossl driver: Fixed authentication mode anon](https://github.com/alorbach/rsyslog-1/commit/e47adc334a946f51a3fa53d51e5ecb48e7375dc8), but I still observed an issue attempting to connect a client without a key and certificate and the referenced commit in fact added client keys and certs that in my view made the test case miss testing clients who don't provide client certs with anon.

### Expected behavior

1. `streamDriver.authMode="anon"` should permit clients to connect without a client cert.
1. The behaviour for `streamDriver.name="gtls"` and `streamDriver.name="ossl"` should be consistent.

Also, the test case written in `sndrcv_tls_ossl_anon_ipv4.sh` should attempt to connect without a client cert.

As per [RFC5424 Client Certficate section](https://tools.ietf.org/html/rfc5246#section-7.4.6):

> If the client does not send any certificates, the server MAY at its discretion either continue the handshake without client authentication [...]

For GNUTLS, 'anon' seemed to imply the above behaviour. The OpenSSL implementation of rsyslog doesn't seem to permit this same behaviour.

### Actual behavior

1. `streamDriver.authMode="anon"` cannot connect without a client cert if `streamDriver.name="ossl"` is set.
1. The behaviour for `streamDriver.name="gtls"` and `streamDriver.name="ossl"` is not consistent with regard to  `streamDriver.authMode="anon"`.

Also, the test case written in `sndrcv_tls_ossl_anon_ipv4.sh` connects *with* a client cert and key hence not testing what happens when a client is anonymous, i.e. has not got a client key or cert.

Notice the line [here](https://github.com/rsyslog/rsyslog/blob/da497b39697d70db2a405bb7aa64e009641d6e6d/tests/sndrcv_tls_ossl_anon_ipv4.sh#L17) for `sndrcv_tls_ossl_anon_ipv4.sh` tests with client cert and key values provided?

```
	defaultNetstreamDriverCertFile="'$srcdir/testsuites/x.509/client-cert.pem'"
	defaultNetstreamDriverKeyFile="'$srcdir/testsuites/x.509/client-key.pem'"
```

### Steps to reproduce the behavior

Config example:

```
module(
  load="imtcp"
  streamDriver.name="gtls"
  #streamDriver.name="ossl"
  streamDriver.mode="1"
  streamDriver.authMode="anon"
)
```

### Environment

- rsyslog version: 8.1907.0
- platform: CentOS 7

**Comment by alorbach** [comment]
Can you please post the error returned by the client? 
Perhaps server and client cannot negotiate a propper anon cipher.

**Comment by rgerhards** [comment]
@JPvRiel note: @alorbach means this test: https://github.com/rsyslog/rsyslog/blob/master/tests/sndrcv_tls_ossl_serveranon_ossl_clientanon.sh

There you can see both client and server configs. This test runs routinely and flawless. Do your problems persist with this config settings?

**Comment by JPvRiel** [comment]
Thanks @alorbach and @rgerhards , I think the issue is that with gnutls, I was able to (ab)use the authMode='anon' to allow clients to be configured without any client cert. This doesn't work for openssl anymore.

Arguably, the ideal would be that the streamDriver openssl auth mode can accept BOTH client's that have certificates and clients that don't. This is known as an Optional Client Certificate request and I'm quite sure the underlying OpenSSL lib does support it. Maybe it needs a wireshark TLS handshake inspection to show how GnuTLS and OpenSSL are behaving differently for the `anon` setting?

**Comment by JPvRiel** [comment]
TL;DR after testing more:

- OpenSSL is not optionally requesting a client cert with `anon` whereas GNUTLS did.
- On latest CentOS 7, there's an openssl version dependency error message, but TLS connections still work.

@alorbach and @rgerhards, please excuse me. I had made a trivial mistake when I first posted the issue. What had happened was I had not installed the openssl dependency into my CentOS test container (but had installed the gnutls one). Noticed this:

```
rsyslogd: could not load module 'lmnsd_ossl', errors: trying to load module /usr/lib64/rsyslog/lmnsd_ossl.so: /usr/lib64/rsyslog/lmnsd_ossl.so: cannot open shared object file: No such file or directory [v8.1907.0 try https://www.rsyslog.com/e/2066 ]
```

I retested now with v8.1908.0 and the rsyslog-openssl rpm package installed. Noticed a new error:

```
rsyslogd: Error: OpenSSL Version to old, SSL_CONF_cmd API  is not supported. [v8.1908.0 try https://www.rsyslog.com/e/2095 ]
```

Nonetheless, the above error, SSL connections without a client cert still work! So perhaps the above needs to be downgraded to a warning. Or perhaps the above is why the openssl stream driver is not requesting an optional client cert?

Via openssl s_client and analysis of packet captures, the behaviour differs as follows:

- GnuTLS syslog TCP TLS service optionally requests a client cert and accepts the client (openssl s_client) not sending any cert. In the packet caputure I see the client sending a sort of empty/nothing type (anon) cert reply.
- OpenSSL doesn't request a client cert at all.

The use-case broken by the OpenSSL behaviour is optionally accepting client certs. With `anon`, OpenSSL forces either no client certs, or always client certs?

I'd like the ability to both support TLS syslog clients who don't have certs and those few clients who do take the time to configure certs.

Also note, TLS settings are globally applied when loading the `imtcp` module, so it's not possible to have 2 separate listeners with different TLS requirements (unlike `imrelp` where TLS settings are per input).

## GnuTLS with `anon`

Tested config extract:

```
module(
  load="imtcp"
  streamDriver.name="gtls"
  #streamDriver.name="ossl"
  streamDriver.mode="1"
  streamDriver.authMode="anon"
)
```

openssl s_client command:

```
openssl s_client -CAfile test/tls_x509/certs/test_ca.cert.pem -msg -state -connect 172.17.0.3:6514 <<< "gnutls test"
```
For s_client `-msg` output, `<<<` is server, and  `>>>` is client.

Notable s_client output (edited for brevity to just show summary TLS messages) and what I observed is:

```
>>> ??? [length 0005]
>>> TLS 1.3, Handshake [length 0120], ClientHello
<<< ??? [length 0005]
<<< TLS 1.3, Handshake [length 0057], ServerHello
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 0479], Certificate
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 014d], ServerKeyExchange
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 3c99], CertificateRequest
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 0004], ServerHelloDone
>>> ??? [length 0005]
>>> TLS 1.2, Handshake [length 0007], Certificate
>>> ??? [length 0005]
>>> TLS 1.2, Handshake [length 0046], ClientKeyExchange
>>> ??? [length 0005]
>>> TLS 1.2, ChangeCipherSpec [length 0001]
>>> ??? [length 0005]
>>> TLS 1.2, Handshake [length 0010], Finished
<<< ??? [length 0005]
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 0010], Finished
>>> ??? [length 0005]
>>> ??? [length 0005]
>>> TLS 1.2, Alert [length 0002], warning close_notify
```

Take note of `<<< TLS 1.2, Handshake [length 3c99], CertificateRequest` followed by `>>> TLS 1.2, Handshake [length 0007], Certificate`

Complete openssl s_client output: [gnutls_s_client.out.txt](https://github.com/rsyslog/rsyslog/files/3556213/gnutls_s_client.out.txt)

From `sudo tcpdump -w gnutls.pcap -i docker0 'port 6514'`, wireshark shows

![gnutls_wireshark_certrequest_seen](https://user-images.githubusercontent.com/9722317/63952308-9de1f780-ca7f-11e9-88c8-8bd2ccde81ad.png)

Capture file: [gnutls.zip](https://github.com/rsyslog/rsyslog/files/3556437/gnutls.zip)

## OpenSSL with `anon`

Took exact same container and config and simply switched streamDriver.name:

```
module(
  load="imtcp"
  #streamDriver.name="gtls"
  streamDriver.name="ossl"
  streamDriver.mode="1"
  streamDriver.authMode="anon"
)
```

Then restarted rsyslog:

openssl s_client command:

```
openssl s_client -CAfile test/tls_x509/certs/test_ca.cert.pem -msg -state -connect 172.17.0.3:6514 <<< "openssl test"
```

Notable s_client output (edited for brevity to just show summary TLS messages) and what I observed is:

```
>>> ??? [length 0005]
>>> TLS 1.3, Handshake [length 0120], ClientHello
<<< ??? [length 0005]
<<< TLS 1.3, Handshake [length 003d], ServerHello
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 0479], Certificate
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 014d], ServerKeyExchange
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 0004], ServerHelloDone
>>> ??? [length 0005]
>>> TLS 1.2, Handshake [length 0046], ClientKeyExchange
>>> ??? [length 0005]
>>> TLS 1.2, ChangeCipherSpec [length 0001]
>>> ??? [length 0005]
>>> TLS 1.2, Handshake [length 0010], Finished
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 00aa], NewSessionTicket
<<< ??? [length 0005]
<<< ??? [length 0005]
<<< TLS 1.2, Handshake [length 0010], Finished
>>> ??? [length 0005]
>>> ??? [length 0005]
>>> TLS 1.2, Alert [length 0002], warning close_notify
```

I don't see any client certificate request, so I assume this implies a client can't use a client cert.

Complete openssl s_client output: [openssl_s_client.out.txt](https://github.com/rsyslog/rsyslog/files/3556248/openssl_s_client.out.txt)

From `sudo tcpdump -w openssl.pcap -i docker0 'port 6514'`, wireshark shows:

![openssl_wireshark_certrequest_NOT_seen](https://user-images.githubusercontent.com/9722317/63952997-c74f5300-ca80-11e9-9171-ba30474523b3.png)

Capture file: [openssl.zip](https://github.com/rsyslog/rsyslog/files/3556435/openssl.zip)

## GnuTLS with `anon` and using optional client cert

Again, to be sure, I captured traffic for GNUTLS to show that a client cert is in fact permitted.

openssl s_client command:

```
openssl s_client -CAfile test/tls_x509/certs/test_ca.cert.pem -cert test/tls_x509/certs/test_syslog_client_centos7.cert.pem -key test/tls_x509/private/test_syslog_client_centos7.key.pem -msg -state -connect 172.17.0.3:6514 <<< "gnutls test with client cert"
```

From the openssl s_client output, I can see the client cert is sent:

```
>>> TLS 1.2, Handshake [length 09d3], Certificate
```

See: [gnutls_s_client_with_client_cert.out.txt](https://github.com/rsyslog/rsyslog/files/3556417/gnutls_s_client_with_client_cert.out.txt)

Wireshark shows the sever requests a client cert listing the test CA as an acceptable authority

![gnutls_wireshark_certrequest_with_test_ca_seen](https://user-images.githubusercontent.com/9722317/63956224-51e68100-ca86-11e9-868c-ed2b6979558e.png)

Wireshark shows the client was able to use it's client cert

![gnutls_client_cert_wireshark_client_cert_optionally_sent](https://user-images.githubusercontent.com/9722317/63956241-5ad75280-ca86-11e9-974e-dfcb91eeafef.png)

Capture file: [gnutls_client_cert.zip](https://github.com/rsyslog/rsyslog/files/3556431/gnutls_client_cert.zip)

## Broader purpose

My view on why it's good to support optional client cert authentication 

- rsyslog TLS could adapt better to complex enterprise networks where some clients are able to send client certs and others not, but still allow clients without client certs to benefit from encrypted communication and authenticate the syslog server at least.
- It'd be great if rsyslog had a property that contained the identify of the client, e.g. `hostauthentication`, similar to `fromhost`, but an authenticated host identity. (needless to say, this should be a separate feature request). Then templates could output that for assurance when the client was identity was authenticated. E.g. it could be the cert fingerprint as used by `permittedPeer` config options.

**Comment by alorbach** [comment]
Regarding this error: 

> rsyslogd: Error: OpenSSL Version to old, SSL_CONF_cmd API  is not supported. [v8.1908.0 try https://www.rsyslog.com/e/2095 ]

It is related to the gnutlsPriorityString configuration parameter which cannot be used on the OpenSSL Version which comes with CentOS7 as it does not support this API. 

Regarding your anon communcation analysis, I am not sure if this is something we can changed on rsyslog side. We are using OpenSSL / GnuTLS API calls to do the TLS Handshake but what happens within these handshakes lies within the API. Its possible that newer OpenSSL Version handle the handshake different.

**Comment by rgerhards** [comment]
> It is related to the gnutlsPriorityString configuration parameter which cannot be used on the OpenSSL Version which comes with CentOS7 as it does not support this API.

I have the impression that this error message is always triggered, even if "gnutlsPriorityString" is not used. @alorbach can you have a look and fix, if so?

**Comment by alorbach** [comment]
If this happens without using gnutlsPriorityString, LogError should just be changed into LogWarning here: 
https://github.com/rsyslog/rsyslog/blob/master/runtime/nsd_ossl.c#L1857

**Comment by JPvRiel** [comment]
Hi @alorbach 

> I am not sure if this is something we can changed on rsyslog side. We are using OpenSSL / 
GnuTLS API calls to do the TLS Handshake but what happens within these handshakes lies within the API

I think it must be settable via the API. For example, the man page for openssl s_server states the following:

> -verify depth, -Verify depth
The verify depth to use. This specifies the maximum length of the client certificate chain and makes the server request a certificate from the client. With the -verify option a certificate is requested but the client does not have to send one, with the -Verify option the client must supply a certificate or an error occurs.

And I've been able to confirm openssl can do optional client certificate requests by testing with the openssl `s_server` and `s_client` apps.

I think that with `anon`, you can make it match GnuTLS behavious if you switch `SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT` to this `SSL_VERIFY_PEER | SSL_VERIFY_CLIENT_ONCE`, which I base on examining the source code of `nsd_ossl.c` vs `s_server.c` in the openssl project  

OpenSSL documents this at [SSL_CTX_set_verify](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify.html)

> SSL_VERIFY_PEER
Server mode: the server sends a client certificate request to the client. The certificate returned (if any) is checked. If the verification process fails, the TLS/SSL handshake is immediately terminated with an alert message containing the reason for the verification failure. The behaviour can be controlled by the additional SSL_VERIFY_FAIL_IF_NO_PEER_CERT, SSL_VERIFY_CLIENT_ONCE and SSL_VERIFY_POST_HANDSHAKE flags.

> SSL_VERIFY_FAIL_IF_NO_PEER_CERT
Server mode: if the client did not return a certificate, the TLS/SSL handshake is immediately terminated with a "handshake failure" alert. This flag must be used together with SSL_VERIFY_PEER.

> SSL_VERIFY_CLIENT_ONCE
Server mode: only request a client certificate once during the connection. Do not ask for a client certificate again during renegotiation or post-authentication if a certificate was requested during the initial handshake. This flag must be used together with SSL_VERIFY_PEER.

## RSyslog `nsd_ossl.c` code

I noticed that in [nsd_ossl.c verification mode](https://github.com/rsyslog/rsyslog/blob/c136aa899b516a5a35767ff3bc511949a4d65a5f/runtime/nsd_ossl.c#L632), a `SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT` value is passed to `SSL_set_verify()`.


## OpenSSL `s_server.c` code

I skimmed the openssl s_client source code. This [line](https://github.com/openssl/openssl/blob/18d307e98ebfe6d376ab039a73dc285190b717a0/apps/s_client.c#L625) was of interest:

```c
 {"verify", OPT_VERIFY, 'p', "Turn on peer certificate verification"}
```
And [OPT_VERIFY](https://github.com/openssl/openssl/blob/18d307e98ebfe6d376ab039a73dc285190b717a0/apps/s_client.c#L1109) lead to:

```c
        case OPT_VERIFY:
            s_server_verify = SSL_VERIFY_PEER | SSL_VERIFY_CLIENT_ONCE;
            verify_args.depth = atoi(opt_arg());
            if (!s_quiet)
                BIO_printf(bio_err, "verify depth is %d\n", verify_args.depth);
            break;
        case OPT_UPPER_V_VERIFY:
            s_server_verify =
                SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT |
                SSL_VERIFY_CLIENT_ONCE;
            verify_args.depth = atoi(opt_arg());
            if (!s_quiet)
                BIO_printf(bio_err,
                           "verify depth is %d, must return a certificate\n",
                           verify_args.depth);
            break;
```

As well as [SSL_get_peer_certificate](https://github.com/openssl/openssl/blob/2f9789f7e7e8f17da32e1cadbbc3c398f99f2d23/apps/s_server.c#L2864):

```c
    peer = SSL_get_peer_certificate(con);
    if (peer != NULL) {
        BIO_printf(bio_s_out, "Client certificate\n");
        PEM_write_bio_X509(bio_s_out, peer);
        dump_cert_text(bio_s_out, peer);
        X509_free(peer);
        peer = NULL;
    }
```

**Comment by alorbach** [comment]
@JPvRiel: 

I reviewed your answer. I have created a testcase in my personal branch here: 
https://github.com/alorbach/rsyslog-1/blob/pr-issue-3837-close/tests/sndrcv_tls_ossl_serveranon_ossl_clientcert.sh

The Server then actually does verify the peer certificate: 
`9132.687089818:main Q:Reg/w0  : nsd_ossl.c: osslPostHandshakeCheck: Debug Version: TLSv1.0 Name: AECDH-AES256-SHA
9132.687092221:main Q:Reg/w0  : nsd_ossl.c: osslChkPeerAuth: Check peer valid[0x7fcae00055d0]
9132.687095036:main Q:Reg/w0  : nsd_ossl.c: osslChkPeerCertValidity: client certificate validation success: ok
`

The behaviour doesn't change even if I add this to CERTANON Mode: 
`SSL_set_verify(pThis->ssl, SSL_VERIFY_PEER|SSL_VERIFY_CLIENT_ONCE, verify_callback);`
